### PR TITLE
Fix various issues reported by Coverity

### PR DESCRIPTION
**Pull request by WOnder93**
This PR contains patches that fix several issues found by a (private) [Coverity](https://scan.coverity.com/) scan.

Note: Synopsys provides free Coverity scans for all open source projects, which can even be launched automatically by Travis CI, but I haven't had time to look into it/set it up.

### CODE REVIEW

**Comment by smuellerDD** [comment]
Thank you very much for the patches.

The patch "check return code of open" however looks strange: open can return 0 as a valid FD. I.e. when stdin is closed and you call open, you get an FD == 0. So, the fix does not seem right.

**Comment by WOnder93** [comment]
You are reading the code wrong, it is the master code that treats 0 as error, thus leading to a possible leak. Hard to see upon quick look, it is, because of the Yoda-style comparison :)

**Comment by smuellerDD** [comment]
Yes, you are right. I guess I need to revert the Yoda-style approach :-)

I started using it because I once introduced a grave bug in another project by assigning instead of comparing. But the current compilers will warn you about that these days. So, no need of Yoda-style.

File: apps/kcapi-hasher.c [hunk 1]
```diff
@@ -706,7 +706,7 @@ static int fipscheck_self(const struct hash_params *params_bin,
 	/* Integrity check of our application. */
 	if (mode == SELFCHECK_CHECK || mode == SELFCHECK_PRINT_SELF) {
 		memset(selfname, 0, sizeof(selfname));
-		selfnamesize = readlink("/proc/self/exe", selfname, BUFSIZE);
+		selfnamesize = readlink("/proc/self/exe", selfname, BUFSIZE - 1);
```
Comment (WOnder93) [inline apps/kcapi-hasher.c]: I just realized the case when readlink() fills the whole buffer is already covered by the check below (and the rest by the memset() call). Please drop this patch when applying. It seems to be a false positive by Coverity (it should check if we check the return value).

**Comment by WOnder93** [comment]
I have dropped one of the patches (the one about `readlink()` and `selfname` in `kcapi-hasher.c`) and added a few more fixes for issues found after rescanning.

**Comment by WOnder93** [comment]
OK, another mistake in the iovec patch... Hopefully all fixed now.

**Comment by smuellerDD** [comment]
Thanks a lot for all the patches.

I have applied them except:

- https://github.com/smuellerDD/libkcapi/pull/65/commits/6266d2e73b88e88f3732326edcc8c72097e06ebf - the reason is that the patch replaces one addition per loop (the advancement of the buffer) with two additions (advancement of the buffer with offset and the advancement of the offset value). I.e. the CPU has to do more work. Which warning did you get from clang?

- due to not considering the mentioned patch, patch https://github.com/smuellerDD/libkcapi/pull/65/commits/ddfdf4282147692a0c6f877db07c1cac80cb117f would not work either. Thus, I deferred it. The patch is definitely needed, thanks for pointing that out. Let us agree on the former issue and then address this one.

**Comment by WOnder93** [comment]
> 6266d2e - the reason is that the patch replaces one addition per loop (the advancement of the buffer) with two additions (advancement of the buffer with offset and the advancement of the offset value). I.e. the CPU has to do more work. Which warning did you get from clang?

If I remove the part after `/* Shut up Clang */`, then I get the following warning:
```
lib/kcapi-kdf.c:394:4: warning: Null pointer passed as an argument to a 'nonnull' parameter
                        memcpy(dst, prk_tmp, dlen);
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

The problem is that if I leave it, then Coverity detects it as unreachable code. I could just mark it as false positive, but I wanted to try to clean it up so that both checkers are happy.

I would argue that the extra addition should be negligible in comparison to the crypto+netlink overhead, but I don't have a proof, so I will not force it. I have now found a simpler solution, I will post it via a new pull request.

> due to not considering the mentioned patch, patch ddfdf42 would not work either. Thus, I deferred it. The patch is definitely needed, thanks for pointing that out. Let us agree on the former issue and then address this one.

OK, I understand.

**Pull request closed without merging**